Remove pkg_config

The pkg_config dependency is causing problems on Windows and on Mac. On Windows the executable doesn't normally exist at all, and on Mac it's not properly set up for R and thus can't find the right directories.

However, in looking over the code in `build.rs`, I don't see a good reason to use pkg_config at all. It's only use is to find a few directories, and whenever you have a working R install R can do that work for you with the `R.home()` function:

    R.home()
    #> [1] "/Library/Frameworks/R.framework/Resources"
    R.home("lib")
    #> [1] "/Library/Frameworks/R.framework/Resources/lib"
    R.home("include")
    #> [1] "/Library/Frameworks/R.framework/Resources/include"

So my recommendation would be to replace the pkg_config calls with a system call to the R executable, something like the following (but from within Rust, not using the bash command line):

    clauswilke ~> R -s -e 'cat(R.home("include"), "\n")'
    /Library/Frameworks/R.framework/Resources/include 

This should be straightforward with the `std::process` module. If desired, multiple directories can be queried with a single call to R via:

    clauswilke ~> R -s -e 'cat(R.home("include"), R.home("lib"), sep = "\n")'
    /Library/Frameworks/R.framework/Resources/include
    /Library/Frameworks/R.framework/Resources/lib

To try this out, I just injected the relevant paths manually, by adding:

```Rust
let r_home = "/Library/Frameworks/R.framework/Resources";
```

and

```Rust
    let include_path = "/Library/Frameworks/R.framework/Resources/include";
    let bindgen_builder = bindgen_builder.clang_arg(format!("-I{}", &include_path));
```

and commenting out all pkg-config related code. `cargo build` now succeeds, but `cargo test` fails with a linking error. (Expand the details to see the exact error message.)

<details>
```
clauswilke ~/github/libR-sys> cargo test
   Compiling libR-sys v0.1.8 (/Users/clauswilke/github/libR-sys)
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-L" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.16fmyq59p2vrdeyp.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.19kpy1qfuuayxvdi.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1ggay8tqa5na8zok.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1nngdqe5xq65b14g.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1owtnb400ji144bc.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1sc0e7toythndd7o.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1t6ajqcv6z6404hq.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.23e1q5018idlordb.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.23fpqq4zzgctdf3o.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.25iq44i9yd3xlkco.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.2fqwwkg0p9ptrnm9.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.2uxxa9msml48r1n4.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.32a814xc0qsxlp2j.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.3a7pnjd9i2oky85p.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.3wvv8ve402z4uuov.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.3zf40r2wrullnd37.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.400x0vx74wtw44wp.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.43cztjfctzvw3br5.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.44dr07o9yv3on1us.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.45te69powc4q14fi.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.4h825qef9o46iial.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.4uxrv9dfy2tk9100.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.57twhzexx99nxlp4.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.hhxg57qq8nqimcc.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.jguwhyr2wks1bua.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.lb0tkqnqra3fm5v.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.mxh5mp2wfy7pms3.rcgu.o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.twrpl2qzjeodqi4.rcgu.o" "-o" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312" "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1egv4m4fwhm1tyna.rcgu.o" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/clauswilke/github/libR-sys/target/debug/deps" "-L" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libtest-40c8f6883f54c8c1.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libterm-e469f66b2fc96a2c.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libgetopts-5d0549d1d51761d5.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunicode_width-ae31bb4ee9e885ea.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_std-1deb66bd98b448b5.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libstd-5aac390c25a3be91.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-8d95e919acff6358.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-b74724130e02bca1.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-0e0d29142d1c9272.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libbacktrace-f1652c422fbf7fc7.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libbacktrace_sys-3d74ad27c9a94aec.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-1255ec28ac7c09c4.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunwind-8753dbda32b8facf.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-a00fd94477b6f959.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liblibc-764e6791ca18d144.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc-a494dce1f057654b.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-2c78f494769c69d7.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcore-47f37590dc106077.rlib" "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-edfd79f9969978bb.rlib" "-lSystem" "-lresolv" "-lc" "-lm"
  = note: Undefined symbols for architecture x86_64:
            "_R_ParseVector", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_VECTOR_ELT", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_Rf_PrintValue", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_R_GlobalEnv", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_R_NilValue", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_Rf_eval", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_TYPEOF", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_R_CStackLimit", referenced from:
                libR_sys::tests::start_R::hd1e98fea0e8b5231 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_REAL", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_Rf_protect", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_Rf_unprotect", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_Rf_initialize_R", referenced from:
                libR_sys::tests::start_R::hd1e98fea0e8b5231 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_Rf_mkString", referenced from:
                libR_sys::tests::test_eval::he8f959a4c30abfb7 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
            "_setup_Rmainloop", referenced from:
                libR_sys::tests::start_R::hd1e98fea0e8b5231 in libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: aborting due to previous error

error: could not compile `libR-sys`.
```
</details>

As far as I can tell, all the missing symbols are from the R library, so that isn't properly being linked. I don't know enough about linking to have a good handle at fixing this problem. Any pointers or things to try would be appreciated.

One other comment: To get this all to run on Mac I had to install a separate version of llvm, because the one that ships with Mac apparently doesn't have `llvm-config` and the bindgen script requires it. This makes me a bit uncomfortable also, because it's another hurdle to get this onto CRAN. Long-term, it may be necessary to distribute pre-generated libR-sys code rather than auto-generate it on the fly.

Hi Claus,

This is exactly why I made libR-sys a separate project.

It may be worth seeing what others (such as the original bingen tutorial)
are doing.

I'm all for removing a dependency like pkg_config from the build if it makes
things easier. It is a shame that LLVM on the Mac is the bottleneck. They
did
fund much of its development by employing Chris Lattner.

I've tried to keep libR-sys as vanilla as possible, but complexity
has crept in.

Andy.



On Wed, Nov 11, 2020 at 4:31 PM Claus Wilke <notifications@github.com>
wrote:

> To try this out, I just injected the relevant paths manually, by adding:
>
> let r_home = "/Library/Frameworks/R.framework/Resources";
>
> and
>
>     let include_path = "/Library/Frameworks/R.framework/Resources/include";
>     let bindgen_builder = bindgen_builder.clang_arg(format!("-I{}", &include_path));
>
> and commenting out all pkg-config related code. cargo build now succeeds,
> but cargo test fails with a linking error. (Expand the details to see the
> exact error message.)
> ``` clauswilke ~/github/libR-sys> cargo test Compiling libR-sys v0.1.8
> (/Users/clauswilke/github/libR-sys) error: linking with `cc` failed: exit
> code: 1 | = note: "cc" "-m64" "-L"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.16fmyq59p2vrdeyp.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.19kpy1qfuuayxvdi.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1ggay8tqa5na8zok.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1nngdqe5xq65b14g.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1owtnb400ji144bc.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1sc0e7toythndd7o.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1t6ajqcv6z6404hq.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.23e1q5018idlordb.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.23fpqq4zzgctdf3o.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.25iq44i9yd3xlkco.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.2fqwwkg0p9ptrnm9.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.2uxxa9msml48r1n4.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.32a814xc0qsxlp2j.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.3a7pnjd9i2oky85p.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.3wvv8ve402z4uuov.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.3zf40r2wrullnd37.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.400x0vx74wtw44wp.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.43cztjfctzvw3br5.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.44dr07o9yv3on1us.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.45te69powc4q14fi.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.4h825qef9o46iial.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.4uxrv9dfy2tk9100.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.57twhzexx99nxlp4.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.hhxg57qq8nqimcc.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.jguwhyr2wks1bua.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.lb0tkqnqra3fm5v.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.mxh5mp2wfy7pms3.rcgu.o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.twrpl2qzjeodqi4.rcgu.o"
> "-o"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312"
> "/Users/clauswilke/github/libR-sys/target/debug/deps/libR_sys-d0f6aed4231ee312.1egv4m4fwhm1tyna.rcgu.o"
> "-Wl,-dead_strip" "-nodefaultlibs" "-L"
> "/Users/clauswilke/github/libR-sys/target/debug/deps" "-L"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libtest-40c8f6883f54c8c1.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libterm-e469f66b2fc96a2c.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libgetopts-5d0549d1d51761d5.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunicode_width-ae31bb4ee9e885ea.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_std-1deb66bd98b448b5.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libstd-5aac390c25a3be91.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-8d95e919acff6358.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-b74724130e02bca1.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-0e0d29142d1c9272.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libbacktrace-f1652c422fbf7fc7.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libbacktrace_sys-3d74ad27c9a94aec.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-1255ec28ac7c09c4.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunwind-8753dbda32b8facf.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-a00fd94477b6f959.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liblibc-764e6791ca18d144.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc-a494dce1f057654b.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-2c78f494769c69d7.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcore-47f37590dc106077.rlib"
> "/Users/clauswilke/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-edfd79f9969978bb.rlib"
> "-lSystem" "-lresolv" "-lc" "-lm" = note: Undefined symbols for
> architecture x86_64: "_R_ParseVector", referenced from:
> libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_VECTOR_ELT", referenced
> from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_Rf_PrintValue",
> referenced from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_R_GlobalEnv",
> referenced from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_R_NilValue", referenced
> from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_Rf_eval", referenced
> from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_TYPEOF", referenced
> from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_R_CStackLimit",
> referenced from: libR_sys::tests::start_R::hd1e98fea0e8b5231 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_REAL", referenced from:
> libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_Rf_protect", referenced
> from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_Rf_unprotect",
> referenced from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_Rf_initialize_R",
> referenced from: libR_sys::tests::start_R::hd1e98fea0e8b5231 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_Rf_mkString",
> referenced from: libR_sys::tests::test_eval::he8f959a4c30abfb7 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o "_setup_Rmainloop",
> referenced from: libR_sys::tests::start_R::hd1e98fea0e8b5231 in
> libR_sys-d0f6aed4231ee312.4wr8w90rfrj4lacz.rcgu.o ld: symbol(s) not found
> for architecture x86_64 clang: error: linker command failed with exit code
> 1 (use -v to see invocation)
>
> error: aborting due to previous error
>
> error: could not compile libR-sys.
>
> </details>
>
> As far as I can tell, all the missing symbols are from the R library, so that isn't properly being linked. I don't know enough about linking to have a good handle at fixing this problem. Any pointers or things to try would be appreciated.
>
> One other comment: To get this all to run on Mac I had to install a separate version of llvm, because the one that ships with Mac apparently doesn't have `llvm-config` and the bindgen script requires it. This makes me a bit uncomfortable also, because it's another hurdle to get this onto CRAN. Long-term, it may be necessary to distribute pre-generated libR-sys code rather than auto-generate it on the fly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/extendr/libR-sys/issues/6#issuecomment-725524023>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAL36XGRNXSGH2SRGVHXRADSPK37DANCNFSM4TRTQCTQ>
> .
>

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "libR-sys"
-version = "0.1.8"
+version = "0.1.9"
 authors = ["andy-thomason <andy@andythomason.com>"]
 edition = "2018"
 description = "Low level bindings to the R programming language."
@@ -12,5 +12,4 @@ documentation = "https://extendr.github.io/libR-sys/master/libR_sys/index.html"
 
 [build-dependencies]
 bindgen = "0.53.2"
-pkg-config = "0.3.17"
 regex = "1"

--- a/README.md
+++ b/README.md
@@ -30,7 +30,9 @@ Set `LIBCLANG_PATH` to the `lib` directory of your `llvm` installation. E.g.,
 ### Windows-specific instructions
 
 Ensure the preferred R binaries, are part of the `PATH`, e.g. `C:/R/R-4.0.2/bin/x64`.
-For information on how to add environment variables on windows, [see here](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.1#saving-changes-to-environment-variables).
+For information on how to add environment variables on Windows, [see here](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.1#saving-changes-to-environment-variables).
+
+Add the mingw toolchains that are used to build R:
 
 ```bash
 rustup target add x86_64-pc-windows-gnu
@@ -60,7 +62,8 @@ pacman -S --noconfirm mingw32/mingw-w64-i686-clang mingw-w64-i686-toolchain
 ```
 
 Add environment variable `LIBCLANG_PATH` with the value pointing to where the
-clang binaries are placed, e.g. with scoop: `%USERPROFILE%\scoop\apps\msys2\current\mingw64\bin`.
+clang binaries are placed. If scoop was used then the path would be:
+`%USERPROFILE%\scoop\apps\msys2\current\mingw64\bin`.
 
 Then from now on, in order to build this, use:
 
@@ -70,7 +73,7 @@ cargo build
 cargo test
 ```
 
-In order to build it without having to enter a `mingw64`/`mingw32` add these
+In order to build it without having to enter a `mingw64`/`mingw32` shell, then add these
 paths to `PATH`:
 
 ```bash

--- a/README.md
+++ b/README.md
@@ -78,8 +78,6 @@ paths to `PATH`:
 %USERPROFILE%\scoop\apps\msys2\current\mingw64\bin # change this to mingw32
 ```
 
-Ensure that `rustc` has the target toolchain that matches R's by:
-
 ### Mac-specific instructions
 
 Install `llvm-config` via [homebrew](https://brew.sh/) with:

--- a/README.md
+++ b/README.md
@@ -9,4 +9,42 @@ Low-level R library bindings
 [API Documentation](https://extendr.github.io/libR-sys/master/libR_sys/index.html)
 
 
+# Installation
 
+To build this library and test that everything is working as expected, run the following two commands in the library's top-level directory:
+
+```
+cargo build
+cargo test
+```
+
+The build script has the following two dependencies:
+
+1. [R.](https://cran.r-project.org/) It needs to be installed and available in the search path.
+2. [libclang.](https://clang.llvm.org/docs/Tooling.html) Depending on your operating system, you may need to set the `LIBCLANG_PATH` environment variable or add `llvm-config` to your search path.
+
+## Linux-specific instructions
+
+Set `LIBCLANG_PATH` to the `lib` directory of your `llvm` installation. E.g., 
+`LIBCLANG_PATH=/usr/lib/llvm-3.9/lib`.
+
+## Windows-specific instructions
+
+... coming soon ...
+
+## Mac-specific instructions
+
+Install `llvm-config` via [homebrew](https://brew.sh/) with:
+```
+brew install llvm
+```
+
+Add it to your search path via:
+```
+echo 'export PATH="/usr/local/opt/llvm/bin:$PATH"' >> ~/.bash_profile
+```
+
+If you want to compile `libR-sys` from within RStudio, you may also have to add the following line to your `.Renviron` file:
+```
+PATH=/usr/local/opt/llvm/bin:${PATH}
+```

--- a/README.md
+++ b/README.md
@@ -20,8 +20,8 @@ cargo test
 
 The build script has the following two dependencies:
 
-1. [R.](https://cran.r-project.org/) It needs to be installed and available in the search path.
-2. [libclang.](https://clang.llvm.org/docs/Tooling.html) Depending on your operating system, you may need to set the `LIBCLANG_PATH` environment variable or add `llvm-config` to your search path.
+1. [R](https://cran.r-project.org/): It needs to be installed and available in the search path.
+2. [libclang](https://clang.llvm.org/docs/Tooling.html): Depending on your operating system, you may need to set the `LIBCLANG_PATH` environment variable or add `llvm-config` to your search path.
 
 ## Linux-specific instructions
 

--- a/README.md
+++ b/README.md
@@ -8,12 +8,11 @@ Low-level R library bindings
 
 [API Documentation](https://extendr.github.io/libR-sys/master/libR_sys/index.html)
 
-
-# Installation
+## Installation
 
 To build this library and test that everything is working as expected, run the following two commands in the library's top-level directory:
 
-```
+```bash
 cargo build
 cargo test
 ```
@@ -23,28 +22,80 @@ The build script has the following two dependencies:
 1. [R](https://cran.r-project.org/): It needs to be installed and available in the search path.
 2. [libclang](https://clang.llvm.org/docs/Tooling.html): Depending on your operating system, you may need to set the `LIBCLANG_PATH` environment variable or add `llvm-config` to your search path.
 
-## Linux-specific instructions
+### Linux-specific instructions
 
-Set `LIBCLANG_PATH` to the `lib` directory of your `llvm` installation. E.g., 
+Set `LIBCLANG_PATH` to the `lib` directory of your `llvm` installation. E.g.,
 `LIBCLANG_PATH=/usr/lib/llvm-3.9/lib`.
 
-## Windows-specific instructions
+### Windows-specific instructions
 
-... coming soon ...
+Ensure that `R_HOME` is an environment variable pointing to the preferred R
+binaries, e.g. `C:/R/R-4.0.2/bin/x64`. Then add `%R_HOME%` to the `PATH`.
 
-## Mac-specific instructions
+```bash
+rustup target add x86_64-pc-windows-gnu
+rustup target add i686-pc-windows-gnu
+```
+
+The default toolchain must be `(stable|beta|nightly)-gnu`, e.g. to set the default
+toolchain to be nightly mingw:
+
+```bash
+rustup default nightly-gnu
+```
+
+Install `MSYS2`. Using scoop it is done by
+
+```bash
+scoop install msys2
+```
+
+To complete the installation, run `msys2` once, then restart the terminal.
+
+Run `msys2` again, and install Clang and mingw-toolchain via
+
+```bash
+pacman -S --noconfirm mingw-w64-x86_64-clang mingw-w64-x86_64-toolchain
+pacman -S --noconfirm mingw32/mingw-w64-i686-clang mingw-w64-i686-toolchain
+```
+
+Add enivornment variable `LIBCLANG_PATH` with the value pointing to where the
+clang binaries are placed, e.g. with scoop: `%USERPROFILE%\scoop\apps\msys2\current\mingw64\bin`.
+
+Then from now on, in order to build this, use:
+
+```bash
+mingw64
+cargo build
+cargo test
+```
+
+In order to build it without having to enter a `mingw64`/`mingw32` add these
+paths to `PATH`:
+
+```bash
+%USERPROFILE%\scoop\apps\msys2\current\usr\bin
+%USERPROFILE%\scoop\apps\msys2\current\mingw64\bin # change this to mingw32
+```
+
+Ensure that `rustc` has the target toolchain that matches R's by:
+
+### Mac-specific instructions
 
 Install `llvm-config` via [homebrew](https://brew.sh/) with:
-```
+
+```bash
 brew install llvm
 ```
 
 Add it to your search path via:
-```
+
+```bash
 echo 'export PATH="/usr/local/opt/llvm/bin:$PATH"' >> ~/.bash_profile
 ```
 
 If you want to compile `libR-sys` from within RStudio, you may also have to add the following line to your `.Renviron` file:
-```
+
+```bash
 PATH=/usr/local/opt/llvm/bin:${PATH}
 ```

--- a/README.md
+++ b/README.md
@@ -29,8 +29,8 @@ Set `LIBCLANG_PATH` to the `lib` directory of your `llvm` installation. E.g.,
 
 ### Windows-specific instructions
 
-Ensure that `R_HOME` is an environment variable pointing to the preferred R
-binaries, e.g. `C:/R/R-4.0.2/bin/x64`. Then add `%R_HOME%` to the `PATH`.
+Ensure the preferred R binaries, are part of the `PATH`, e.g. `C:/R/R-4.0.2/bin/x64`.
+For information on how to add environment variables on windows, [see here](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.1#saving-changes-to-environment-variables).
 
 ```bash
 rustup target add x86_64-pc-windows-gnu
@@ -59,7 +59,7 @@ pacman -S --noconfirm mingw-w64-x86_64-clang mingw-w64-x86_64-toolchain
 pacman -S --noconfirm mingw32/mingw-w64-i686-clang mingw-w64-i686-toolchain
 ```
 
-Add enivornment variable `LIBCLANG_PATH` with the value pointing to where the
+Add environment variable `LIBCLANG_PATH` with the value pointing to where the
 clang binaries are placed, e.g. with scoop: `%USERPROFILE%\scoop\apps\msys2\current\mingw64\bin`.
 
 Then from now on, in order to build this, use:

--- a/build.rs
+++ b/build.rs
@@ -1,47 +1,65 @@
 extern crate bindgen;
 
-use std::env;
-use std::path::PathBuf;
+use std::{env, io, io::Error, io::ErrorKind, process::Command, process::exit, path::PathBuf};
 use regex::Regex;
 
 struct InstallationPaths {
-    config: Option<String>,
-    library: Option<String>
+    r_home: String,
+    include: String,
+    library: String
 }
 
-fn paths() -> InstallationPaths {
-    if let Ok(r_home) = env::var("R_HOME") {
-        InstallationPaths {
-            config: Some(format!("{}/../pkgconfig", r_home)),
-            library: Some(format!("{}/lib", r_home))
-        }
-    } else {
-        InstallationPaths {
-            config: None,
-            library: None
-        }
-    }
+fn probe_r_paths() -> io::Result<InstallationPaths> {
+    let rout = Command::new("R")
+        .args(&["-s", "-e", r#"cat(R.home(), R.home('include'), R.home('lib'), sep = '\n')"#])
+        .output()?;
+
+    let rout = String::from_utf8_lossy(&rout.stdout);
+    let mut lines = rout.lines();
+
+    let r_home = match lines.next() {
+        Some(line) => line.to_string(),
+        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R home"))
+    };
+
+    let include = match lines.next() {
+        Some(line) => line.to_string(),
+        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R include"))
+    };
+
+    let library = match lines.next() {
+        Some(line) => line.to_string(),
+        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R library"))
+    };
+
+    Ok(InstallationPaths {
+        r_home: r_home,
+        include: include,
+        library: library,
+    })
 }
 
 fn main() {
-    let details = paths();
+    let details = probe_r_paths();
 
-    if let Some(v) = details.config {
-        env::set_var("PKG_CONFIG_PATH", v);
-        env::set_var("PKG_CONFIG_ALLOW_CROSS", "1");
-    }
+    let details = match details {
+        Ok(result) => result,
+        Err(error) => {
+            println!("Problem locating local R instal: {:?}", error);
+            exit(1);
+        }
+    };
 
-    if let Some(v) = details.library {
-        env::set_var("LD_LIBRARY_PATH", v);
-    }
+    // is this required?
+    //env::set_var("LD_LIBRARY_PATH", &details.library);
 
-    let r_lib = pkg_config::probe_library("libR").unwrap();
-    let r_home = pkg_config::get_variable("libR", "rhome").unwrap();
-    println!("cargo:rustc-env=R_HOME={}", r_home);
-    println!("cargo:r_home={}", r_home); // Becomes DEP_R_R_HOME for clients
+    println!("cargo:rustc-env=R_HOME={}", &details.r_home);
+    println!("cargo:r_home={}", &details.r_home); // Becomes DEP_R_R_HOME for clients
+    // make sure cargo links properly against library
+    println!("cargo:rustc-link-search={}", &details.library);
+    println!("cargo:rustc-link-lib=dylib=R");
 
     println!("cargo:rerun-if-changed=build.rs");
-
     println!("cargo:rerun-if-changed=wrapper.h");
 
     // The bindgen::Builder is the main entry point
@@ -62,9 +80,7 @@ fn main() {
         .parse_callbacks(Box::new(bindgen::CargoCallbacks));
 
     // Point to the correct headers
-    let bindgen_builder = r_lib.include_paths
-        .iter()
-        .fold(bindgen_builder, |bb, path| bb.clang_arg(format!("-I{}", path.to_str().unwrap())));
+    let bindgen_builder = bindgen_builder.clang_arg(format!("-I{}", &details.include));
 
     // Finish the builder and generate the bindings.
     let bindings = bindgen_builder
@@ -73,10 +89,6 @@ fn main() {
         .expect("Unable to generate bindings");
 
     // Extract the version number from the R headers.
-    // We could get this from pkgconfig instead, but there
-    // is less ambiguity about the format this way, and will
-    // still work if we allow builds without pkgconfig in the
-    // future.
     let version_matcher = Regex::new(r"pub const R_VERSION ?: ?u32 = (\d+)").unwrap();
     if let Some(version) = version_matcher.captures(bindings.to_string().as_str()) {
         let version = version

--- a/build.rs
+++ b/build.rs
@@ -1,17 +1,25 @@
 extern crate bindgen;
 
-use std::{env, io, io::Error, io::ErrorKind, process::Command, process::exit, path::PathBuf};
 use regex::Regex;
+use std::{env, io, io::Error, io::ErrorKind, path::PathBuf, process::exit, process::Command};
 
 struct InstallationPaths {
     r_home: String,
     include: String,
-    library: String
+    library: String,
 }
 
 fn probe_r_paths() -> io::Result<InstallationPaths> {
     let rout = Command::new("R")
-        .args(&["-s", "-e", r#"cat(R.home(), R.home('include'), R.home('lib'), sep = '\n')"#])
+        .args(&[
+            "-s",
+            "-e",
+            if cfg!(target_os = "windows") {
+                r#"cat(R.home(), R.home('include'), R.home('bin'), sep = '\n')"#
+            } else {
+                r#"cat(R.home(), R.home('include'), R.home('lib'), sep = '\n')"#
+            }
+        ])
         .output()?;
 
     let rout = String::from_utf8_lossy(&rout.stdout);
@@ -19,23 +27,23 @@ fn probe_r_paths() -> io::Result<InstallationPaths> {
 
     let r_home = match lines.next() {
         Some(line) => line.to_string(),
-        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R home"))
+        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R home")),
     };
 
     let include = match lines.next() {
         Some(line) => line.to_string(),
-        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R include"))
+        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R include")),
     };
 
     let library = match lines.next() {
         Some(line) => line.to_string(),
-        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R library"))
+        _ => return Err(Error::new(ErrorKind::Other, "Cannot find R library")),
     };
 
     Ok(InstallationPaths {
-        r_home: r_home,
-        include: include,
-        library: library,
+        r_home,
+        include,
+        library,
     })
 }
 
@@ -79,8 +87,12 @@ fn main() {
         // included header files changed.
         .parse_callbacks(Box::new(bindgen::CargoCallbacks));
 
+        // println!("TARGET: {}",cargo_env("TARGET"));
     // Point to the correct headers
-    let bindgen_builder = bindgen_builder.clang_arg(format!("-I{}", &details.include));
+    let bindgen_builder = bindgen_builder.clang_args(&[
+        format!("-I{}", &details.include),
+        format!("--target={}", std::env::var("TARGET").expect("Could not get the target triple"))
+    ]);
 
     // Finish the builder and generate the bindings.
     let bindings = bindgen_builder
@@ -91,9 +103,7 @@ fn main() {
     // Extract the version number from the R headers.
     let version_matcher = Regex::new(r"pub const R_VERSION ?: ?u32 = (\d+)").unwrap();
     if let Some(version) = version_matcher.captures(bindings.to_string().as_str()) {
-        let version = version
-            .get(1).unwrap().as_str()
-            .parse::<u32>().unwrap();
+        let version = version.get(1).unwrap().as_str().parse::<u32>().unwrap();
         println!("cargo:r_version={}", version);
     } else {
         panic!("failed to find R_VERSION");

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -81,7 +81,9 @@ mod tests {
             // Maybe 8MB is a bit small.
             // eprintln!("R_CStackLimit={:016x}", R_CStackLimit);
 
-            R_CStackLimit = usize::max_value();
+            if cfg!(not(target_os = "windows")) {
+                R_CStackLimit = usize::max_value();
+            }
 
             setup_Rmainloop();
         }

--- a/wrapper.h
+++ b/wrapper.h
@@ -1,14 +1,21 @@
 
 
 // Define this for R_CStackLimit
-//#define HAVE_UINTPTR_T
+// #define HAVE_UINTPTR_T
 #define CSTACK_DEFNS
 
 // Currently, I'm adding these on as-needed basis
 // but we may simply throw the whole lot in in the future.
-#include <Rinterface.h>
 #include <Rinternals.h>
+
+#ifndef _WIN32
+  #define R_INTERFACE_PTRS
+  #include <Rinterface.h>
+#else
+    extern uintptr_t R_CStackLimit; /* C stack limit */
+#endif
 #include <Rembedded.h>
+
 // todo: what is this #define?
 #ifdef MAC_STUFF
   #include <R/Rversion.h>

